src/headers.js

```diff
@@ -1,33 +1,47 @@
 'use strict'
 
 const { types } = require('util')
+const { validateHeaderName, validateHeaderValue } = require('http')
 
-const { normalizeAndValidateHeaderName, normalizeAndValidateHeaderArguments } = require('./utils')
+const { kHeaders } = require('./symbols')
 
-const kHeaders = Symbol('headers')
+function normalizeAndValidateHeaderName (name) {
+  const normalizedHeaderName = name.toLowerCase()
+  validateHeaderName(normalizedHeaderName)
+  return normalizedHeaderName
+}
+
+function normalizeAndValidateHeaderValue (name, value) {
+  const normalizedHeaderName = normalizeAndValidateHeaderName(name)
+  const normalizedHeaderValue = value.replace(/^[\n\t\r\x20]+|[\n\t\r\x20]+$/g, '')
+  validateHeaderValue(normalizedHeaderName, normalizedHeaderValue)
+  return [normalizedHeaderName, normalizedHeaderValue]
+}
+
+function fill (headers, object) {
+  if (Array.isArray(object)) {
+    for (let i = 0, header = object[0]; i < object.length; i++, header = object[i]) {
+      if (header.length !== 2) throw TypeError('header entry must be of length two')
+      headers.append(header[0], header[1])
+    }
+  } else if (!types.isBoxedPrimitive(object)) {
+    for (const [name, value] of Object.entries(object)) {
+      headers.append(name, value)
+    }
+  }
+}
 
 class Headers {
   constructor (init) {
     this[kHeaders] = new Map()
 
     if (init && typeof init === 'object') {
-      if (Array.isArray(init)) {
-        for (let i = 0, header = init[0]; i < init.length; i++, header = init[i]) {
-          if (header.length !== 2) throw TypeError('header entry must be of length two')
-          this.append(header[0], header[1])
-        }
-      } else if (kHeaders in init) {
-        this[kHeaders] = new Map(init[kHeaders])
```

Comment:
This seems regressive performance wise - if these entries have already passed through the `normalizeAndValidate..` append flow.

Reply:
Hmm good point. I'm worry of allowing someone to pass a Headers instance directly to the Headers class as that is against spec; but, since I've moved this to an independent `fill` method maybe its okay

Reply:
fixed in latest commit

Reply:
> as that is against spec

Using a Headers object as HeadersInit **is allowed** since a Headers object can be represented as an IDL `sequence<sequence<ByteString>>`

Reply:
ah i see, so the fill operation does allow us to pass a Headers instance. In the spec, that operation still wants to iterate and call `.append`, but your original solution avoids that unnecessary re-validation 👍 got it